lightning/src/ln/channelmanager.rs

```diff
@@ -2897,6 +2905,33 @@ macro_rules! handle_error {
 /// [`ChannelMonitor`]/channel funding transaction) to begin with.
 macro_rules! update_maps_on_chan_removal {
 	($self: expr, $peer_state: expr, $channel_context: expr) => {{
+		let chan_id = $channel_context.channel_id();
+
+		// If there's a possibility that we need to generate further monitor updates for this
+		// channel, we need to store the last update_id of it. However, we don't want to insert
+		// into the map (which prevents the `PeerState` from being cleaned up) for channels that
+		// never even got confirmations (which would open us up to DoS attacks).
+		let mut update_id = $channel_context.get_latest_monitor_update_id();
+		if $channel_context.get_funding_tx_confirmation_height().is_some() || $channel_context.minimum_depth() == Some(0) || update_id > 1 {
+			// There may be some pending background events which we have to ignore when setting the
+			// latest update ID.
+			for event in $self.pending_background_events.lock().unwrap().iter() {
+				match event {
+					BackgroundEvent::MonitorUpdateRegeneratedOnStartup { counterparty_node_id, channel_id, update, .. } => {
+						if *channel_id == $channel_context.channel_id() && *counterparty_node_id == $channel_context.get_counterparty_node_id() {
+							update_id = cmp::min(update_id, update.update_id - 1);
+						}
+					},
+					BackgroundEvent::ClosedMonitorUpdateRegeneratedOnStartup(..) => {
+						// This is only generated for very old channels which were already closed
+						// on startup, so it should never be present for a channel that is closing
+						// here.
+					},
+					BackgroundEvent::MonitorUpdatesComplete { .. } => {},
+				}
+			}
+			$peer_state.closed_channel_monitor_update_ids.insert(chan_id, update_id);
```

Comment:
I'm getting a test failure with this diff:
```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 5923686b4..f99afc41e 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -1305,7 +1305,7 @@ pub(super) struct PeerState<SP: Deref> where SP::Target: SignerProvider {
        ///
        /// If there are any updates pending in [`Self::in_flight_monitor_updates`] this will contain
        /// the highest `update_id` of all the pending in-flight updates.
-       closed_channel_monitor_update_ids: BTreeMap<ChannelId, u64>,
+       pub(crate) closed_channel_monitor_update_ids: BTreeMap<ChannelId, u64>,
        /// The peer is currently connected (i.e. we've seen a
        /// [`ChannelMessageHandler::peer_connected`] and no corresponding
        /// [`ChannelMessageHandler::peer_disconnected`].
diff --git a/lightning/src/ln/monitor_tests.rs b/lightning/src/ln/monitor_tests.rs
index 99ad31d35..02d713039 100644
--- a/lightning/src/ln/monitor_tests.rs
+++ b/lightning/src/ln/monitor_tests.rs
@@ -211,6 +211,11 @@ fn archive_fully_resolved_monitors() {
                OutPoint { txid: funding_tx.compute_txid(), index: 0 },
                funding_tx.output[0].script_pubkey.clone()
        );
+       let per_peer_state = nodes[0].node.per_peer_state.read().unwrap();
+       if let Some(peer_state_mtx) = per_peer_state.get(&nodes[1].node.get_our_node_id()) {
+               let mut peer_state = peer_state_mtx.lock().unwrap();
+               assert!(peer_state.closed_channel_monitor_update_ids.is_empty());
+       } else { panic!() }
 }

 fn do_chanmon_claim_value_coop_close(anchors: bool) {
```
I thought we should be pruning from `closed_channel_monitor_update_ids` after a monitor is archived?